### ISSUE TITLE
Possible collision of different functions

### ISSUE THREAD

**Issue by suquark** [ray-project/ray#874]
This code shows that different functions can collide in ray:

```python
class A(object):
  def func(self):
    return inspect.getframeinfo(inspect.currentframe()).lineno

class B(object):
  def func(self):
    return inspect.getframeinfo(inspect.currentframe()).lineno

t1 = ray.remote(A.func)
t2 = ray.remote(B.func)

print(ray.get(t1.remote(None)))
print(ray.get(t2.remote(None)))
```

This happens because in `worker.py`, the signature of function is calculated like this:

```python
 def remote_function_decorator(func, function_properties):
   func_name = "{}.{}".format(func.__module__, func.__name__)
   if func_id is None:
     function_id = compute_function_id(func_name, func)
```

So there's nothing about which class the function stays in.

**Comment by robertnishihara** [comment]
Good catch! It looks like in this case both remote functions are identified by the name `__main__.func` and hence have the same function IDs. It seems like this is a problem that would only occur when a method of a class is decorated.

I haven't really thought much about that use case. Do you have a situation where decorating a method of a class like that is a natural thing to do?

One workaround that works is to do the following.

```python
import ray
import inspect

ray.init()

class A(object):
    def func(self):
        return inspect.getframeinfo(inspect.currentframe()).lineno

class B(object):
    def func(self):
        return inspect.getframeinfo(inspect.currentframe()).lineno

@ray.remote
def t1():
    return A.func(None)

@ray.remote
def t2():
    return B.func(None)

print(ray.get(t1.remote()))
print(ray.get(t2.remote()))
```

**Comment by suquark** [comment]
Oh, I just have found a similar case in my project (and would cause bugs) recently. So when I noticed those code in ray I realized that they may cause potential bugs too :-)

My solution is using `__qualname__` instead of `name` (though it hardly works for decorators and only supported by python3.x)

You are right, it should not be a serious situation for ray. Most times `func.__module__+func.__name__` is enough.
